Fix #19167: Allow reviewers to undo most recent submitted translation review.

Ash-2k3 opened the pull request:
## Overview

1. This PR fixes https://github.com/oppia/oppia/issues/19167
2. This PR does the following: This PR changes the review workflow for translation reviewers. When launched, reviewers will be able to undo their most recent submitted review within a timeframe for 30 seconds. We are achieving this by queuing the suggestion object and making a network call to commit the suggestion either when timer expires or user triggers dismiss action.

Feature Flag for this feature: https://github.com/oppia/oppia/pull/19946

## Essential Checklist

- [ ] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...", followed by a short, clear summary of the changes.
- [ ] I have followed the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- [ ] I have assigned the correct reviewers to this PR (or will leave a comment with the phrase "@{{reviewer_username}} PTAL" if I don't have permissions to assign reviewers directly).
- [ ] The linter/Karma presubmit checks pass on my local machine, and my PR follows the [coding style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide)).
- [ ] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)


## Proof that changes are correct

https://github.com/oppia/oppia/assets/98253080/f0af943b-0680-4d8a-a241-93602d5c3395

#### Proof of changes on mobile phone

https://github.com/oppia/oppia/assets/98253080/8dc3976f-3ce7-4c9a-b74d-2406f946fb47

## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

Ash-2k3 committed:
Implement Undo feature

Ash-2k3 committed:
Add custom snackbar

Ash-2k3 committed:
Fix non null assertion error

Ash-2k3 committed:
Fix lint errors

Ash-2k3 committed:
Fix CI

Ash-2k3 committed:
Fix CI

Ash-2k3 committed:
Fix CI

Ash-2k3 committed:
Merge branch 'develop' into undo_translation_status

Ash-2k3 committed:
Fix CI

Ash-2k3 committed:
Fix CI

Ash-2k3 committed:
Fix CI

Ash-2k3 committed:
Fix conflicts

Ash-2k3 committed:
Add unit tests for deleted tests

Ash-2k3 commented:
Hi @chris7716, @kevintab95, PTAL at this PR. 

Note: The changes of this PR were already approved  in https://github.com/oppia/oppia/pull/18831, but it was held long due to consistent failure in frontend checks. This PR introduces the same changes but with a different approach to mock custom snack bars. (I had to close initial PR as there were lot of ambiguous changes)

Ash-2k3 committed:
Merge branch 'develop' into undo_translation_status

Ash-2k3 committed:
Update translation-suggestion-review-modal.component.ts

Ash-2k3 committed:
Fix CI checks

Ash-2k3 committed:
Merge branch 'develop' into undo_translation_status

Ash-2k3 committed:
Feature gate the undo feature

Ash-2k3 commented:
Proof that it has been gated behind feature flag:


https://github.com/oppia/oppia/assets/98253080/7de59dab-176e-4099-a12d-39739ac2be64

Ash-2k3 committed:
Fix flaky behaviour of unit tests

chris7716 commented on core/templates/pages/contributor-dashboard-page/custom-snackbar/undo-snackbar.component.html:
```diff
@@ -0,0 +1,35 @@
+<div class="snackbar-container">
```
I am not sure whether this snackbar component should be inside contributor dashboard page or core/components directory. @seanlip WDYT?

seanlip replied:
I would put it in core/templates/components since it has nothing to do with the contributor dashboard page intrinsically. Good catch!

seanlip replied:
@Ash-2k3 You didn't respond to this comment.

Ash-2k3 replied:
Done, Thanks!

chris7716 commented on core/templates/pages/contributor-dashboard-page/custom-snackbar/undo-snackbar.component.spec.ts:
```diff
@@ -0,0 +1,59 @@
+// Copyright 2023 The Oppia Authors. All Rights Reserved.
```
Update headers to match current year. Please update any other phrases accordingly. Ditto elsewhere.

Ash-2k3 replied:
Done, Thanks! -- Btw should we change dates for files which I haven't created but modified ?, e.g translation-suggestion-review-modal.component.ts

chris7716 commented on core/templates/pages/contributor-dashboard-page/modal-templates/translation-suggestion-review-modal.component.spec.ts:
```diff
@@ -1095,139 +2124,215 @@ describe('Translation Suggestion Review Modal Component', function () {
+      status: 'status',
+      target_type: 'target_type',
+    };
+    const suggestion3Obsolete = {
```
Can we have meaningful names? Ditto elsewhere.

Ash-2k3 replied:
Done, named it to obsoleteSuggestion. [Commit: ](https://github.com/oppia/oppia/pull/19954/commits/4915cede39b71c0aab95ed9a6151c22fea08d9c5)

chris7716 commented on core/templates/pages/contributor-dashboard-page/modal-templates/translation-suggestion-review-modal.component.spec.ts:
```diff
@@ -1095,139 +2124,215 @@ describe('Translation Suggestion Review Modal Component', function () {
     }));
   });
 
-  describe(
```
Why are we removing these?

Ash-2k3 replied:
I am not removing any existing tests here, I have introduced two new describe blocks and also renamed two of the existing ones.

The new ones have suffix as `when flag CdAllowUndoingTranslationReview is enabled` and two existing ones which I renamed have suffix `flag CdAllowUndoingTranslationReview is disabled`.

The reason, I did this was to make sure that clean up of the feature flag remains simple once everything is tested in feature testing.

chris7716 reviewed (changes requested):
Thanks! Took a pass.

Ash-2k3 committed:
Address Review comments

Ash-2k3 commented:
Thanks @chris7716, addressed your comments, PTAL.

kevintab95 reviewed (approved):
LGTM for 2 codeowner files. Thanks @Ash-2k3!

Ash-2k3 committed:
Address review comments

Ash-2k3 committed:
Address review comments

Ash-2k3 committed:
Merge branch 'develop' into undo_translation_status

Ash-2k3 committed:
Update shared-component.module.ts

Ash-2k3 committed:
Add custom snack bars to codeowner files

chris7716 reviewed (commented):
One more comment. Do we need to gate this feature behind a feature flag?

Ash-2k3 commented:
> One more comment. Do we need to gate this feature behind a feature flag?

I believe yes, it's safer for us to gate it behind feature flag and thoroughly test it in test server before releasing into production. Reason is, if we missed any corner case in local testing it might totally block the review process as in this PR, we are changing the core logic of reviewing translations.

chris7716 commented:
> > One more comment. Do we need to gate this feature behind a feature flag?
> 
> I believe yes, it's safer to gate it behind feature flag and thoroughly test it in test server before releasing into production. Reason is, if we missed any corner case in local testing it might totally block the review process as in this PR, we are changing the workflow of reviewing translations.

If we are implementing a feature flag for this feature, I think we should implement in this PR. But I don't see any changes related to them?

Ash-2k3 commented:
@chris7716, I added a feature flag for this in a separate PR https://github.com/oppia/oppia/pull/19946. I have updated the PR description with this link too, Thanks!

Ash-2k3 committed:
Merge branch 'develop' into undo_translation_status
